src/clients/portable/mgrsketch/mgrsketch.1
.\"{{{}}}
.\"{{{  Title
.TH MGRSKETCH "1"
.\"}}}
.\"{{{  Name
.SH NAME
mgrsketch \- a mgr windowing system sketching program
.\"}}}
.\"{{{  Synopsis
.SH SYNOPSIS
.B mgrsketch
.\"}}}
.\"{{{  Description
.SH DESCRIPTION
\fIMgrsketch\fR is a sketching program that runs on the mgr windowing
system.  Lines, freehand curves, and text are drawn on a window using
the mouse and the keyboard.  Images may be saved as mgr bitmap files,
and may be loaded and edited as well.  The operation of the program is
controlled by a menu that is activated by the middle mouse button.  For
some menu items a dialog box pops up asking for supplementary
information.  The right mouse button is used to draw lines on the
window.  Depressing this button starts the drawing activity, while
lifting terminates it.
.LP
The individual menu entries have the following actions:
.\"{{{  Line
.SS Line
Pressing the right button starts a straight line at the current location
of the mouse cursor.  Move the mouse cursor to the desired location of
the end of the line and release the button.  The line will appear.
.\"}}}
.\"{{{  X-line, Y-line
.SS "X-line, Y-line"
These work like \fBline\fR, except that the resulting line will not end
at the mouse button release point, but will either be horizontal or
vertical.
.\"}}}
.\"{{{  Box
.SS "Box"
This works like \fBline\fR, except that a box will appear with diagonal
corners located at the mouse press and release points.
.\"}}}
.\"{{{  Arrowhead
.SS "Arrowhead"
This sets the length in pixels of the arrowhead drawn at the end of a
line, x-line, or y-line.  A length of zero (the default) means that no
arrowhead is drawn.
.\"}}}
.\"{{{  Circle
.SS "Circle"
This makes a circle with the center at the mouse press position and the
mouse release position defining a point on the circle.
.\"}}}
.\"{{{  Paint
.SS "Paint"
This command allows freehand sketches to be made.  The mouse cursor
draws when the right button is depressed, and stops drawing when it is
released.
.\"}}}
.\"{{{  Brushsize
.SS "Brushsize"
This command allows the x and y dimensions (in pixels) of the square
paintbrush that draws all lines to be set.  Enforced limits are 1 to 12
pixels.
.\"}}}
.\"{{{  Text
.SS "Text"
After selecting this command, depress and release the right mouse button
at the location where text is to begin.  Type the desired text at the
keyboard and terminate with a carriage return.
.IP "Warning"
If \fBtext\fR is selected, the right mouse button is pushed, and another
menu item is immediately selected, strange messages will appear as text
on the image.  Avoid this by typing carriage return to terminate an
unwanted text entry before doing anything else.
.sp
Typing anything on the keyboard when text is not expected will confuse
\fImgrsketch\fR when the next menu item is selected.  Reselecting the
desired menu item will then work, however.
.\"}}}
.\"{{{  New font
.SS "New font"
This selects one of mgr's loaded fonts (0\(en14).  A dialog box allows
you to type in the desired font.
.\"}}}
.\"{{{  Erase
.SS "Erase"
This command allows you to erase with the mouse cursor when the right
button is held down.  The eraser size is the same as the size of the
paintbrush.
.\"}}}
.\"{{{  Load file
.SS "Load file"
This command allows you to load a bitmap file in mgr bitmap file format.
A dialog box asks for the file name.  The bitmap image is immediately
displayed on the window.  If the file can't be found, or if it isn't an
mgr bitmap file, a dialog box will complain.
.IP "Warning"
Loading a bitmap bigger than the current window will result in the
bitmap being truncated.
.sp
The environment variable "PWD", containing the current working
directory, is expected to exist.  If it doesn't, mgrsketch may have
trouble finding an image to load.
.\"}}}
.\"{{{  Save file
.SS "Save file"
This command allows you to save the image in the window to mgr bitmap
file format.  A dialog box asks for a file name, or if a previous file
has been loaded, asks if it is OK to overwrite that file with the
current image.
.IP "Warning:"
When saving a file, it is first written to /tmp by mgr, and then moved 2
seconds later by mgrsketch to the specified location.  This is done
since mgr runs suid root, and may not be able to write on a NFS-mounted
file system.  However, this opens two additional error possibilities: 1)
If for some reason mgr is very slow to write the file, mgrsketch will
try to move it before it is written.  2)  For some UNIX systems, it may
not be allowed to "mv" a file across file system boundaries.
.\"}}}
.\"{{{  Quit
.SS "Quit"
This terminates the program.  Control-C has the same effect.  Note that
the current image is not saved unless you do it explicitly with the
above command.
.\"}}}
.\"}}}
.\"{{{  Author
.SH AUTHOR
David J. Raymond, Physics Department,
.br
New Mexico Institute of Mining and Technology,
.br
Socorro, NM 87801 USA
.br
raymond@kestrel.nmt.edu
.\"}}}


font/History.txt
Ver 1.1
-------
First public release.

Ver 1.2
-------
1) Greek fonts by Angelos Karageorgiou.
2) Hebrew script and others by Aaron Schmiedel (AFS Ltd.)
3) TeX/Math/Symbols font by Dov Grobgeld.

Ver 1.3
-------
1) Hercules plus card fonts. Debugged by Yossi Gil
2) Runic and other fonts "egafon" archive. Communicated by Dov Grobgeld.
3) Persian fonts from the "aria.zip" archive.

Ver 1.4
-------
1) Esperanto font
2) Armenian font
3) Turkish/Greek/Yugoslavian/Icelandic code pages.
4) Code pages at height 19
5) Arabic fonts.
6) Arabic code page fonts.
7) Updated the persian.f14 font
8) Greek fonts
9) Persian-English-Arabic-Urdu font (APEAUS)
10) Thai font

Ver 1.5
-------
1) Miguel Farah very readable fonts. I can attest that these fonts 
indeed very eye pleasant when high text resolutions are used.


src/clients/portable/mgrsketch/mgrsketch.c
/* mgrsketch -- draw a picture on the screen with mgr windowing system running.
 * David J. Raymond, Physics Department
 * New Mexico Institute of Mining and Technology
 * Socorro, NM 87801
 * raymond@kestrel.nmt.edu
 * $Id: mgrsketch.c,v 1.4 1993/04/24 14:08:55 dave Exp dave $
 */
#include <mgr/mgr.h>
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <signal.h>
#include <string.h>

/* default window dimensions */
#define X0 0
#define Y0 0
#define DX 600
#define DY 400
#define BORDER 12

/* offset to tip of cursor arrow */
#define OFFSET 5

/* text offsets */
#define XTOFF 4
#define YTOFF 4

/* paintbrush size */
#define PSIZE 12

/* maximum number of mgr fonts loadable */
#define MAXNFONTS 15

/* maximum length of line */
#define LINESIZE 256

/* define states */
#define LINE 1
#define ERASE 2
#define TEXT 3
#define XLINE 5
#define YLINE 6
#define BOX 7
#define PAINT 8
#define CIRCLE 9

#define MENU_COUNT (sizeof(menu)/sizeof(struct menu_entry))

#define NCHORDS 36
#define RADIAN 57.2958

char buffer[LINESIZE + 1];			/* buffer for input from mgr server */
char cline[LINESIZE + 1];			/* a line buffer */
char fname[LINESIZE + 1];			/* file name */
char fpath[LINESIZE + 1];			/* pathname */
char fpwd[LINESIZE + 1];			/* current working directory */
int mdown = 0;					/* state of mouse */
int state = LINE;				/* state of system */
int x = 0,y = 0;				/* current cursor position */
int oldx = 0,oldy = 0;				/* previous cursor position */
int xx1,yy1,xx2,yy2;				/* arrowhead points */
int nx,ny,nd;					/* size of loaded bitmap */
int xbrush = 1;					/* paintbrush size */
int xarrow = 0;					/* arrow size */
int cfont = 0;					/* current font */
float fdx,fdy,dangle,angle,radius;		/* stuff for circle draw */
int loop,dx,dy,odx,ody;				/* more stuff for circle draw */
struct menu_entry menu[] = {
	{"line","line\n",},
	{"x-line","xline\n",},
	{"y-line","yline\n",},
	{"box","box\n",},
	{"arrowhead","arrow\n",},
	{"circle","circle\n",},
	{"paint","paint\n",},
	{"brushsize","brushsize\n",},
	{"text","text\r",},
	{"new font","newfont\n",},
	{"erase","erase\n",},
	{"load file","load\n",},
	{"save file","save\n",},
	{"quit","quit\n",},
};
FILE *fileptr;					/* pointer to load and save files */
void cleanup();					/* clean up after interrupt */
char *dbox();					/* dialog box */
void fnclean();
void fatline();
void arrowline();
void move();

int
main()
{

/* set things up */
	m_setup(M_FLUSH);
	m_push(P_POSITION|P_MENU|P_EVENT);
	menu_load(1,MENU_COUNT,menu);
	m_selectmenu(1);
	m_clear();
	m_shapewindow(X0,Y0,DX + BORDER,DY + BORDER);
	m_setmode(M_ABS);
	m_setevent(BUTTON_1,"rbutdown\n");
	m_setevent(BUTTON_1U,"rbutup\n");
	m_setevent(DESTROY,"destroy\n");
	m_setnoecho();
	m_setcursor(CS_INVIS);
	signal(SIGINT,cleanup);
	m_func(BIT_CLR);
	m_bitwriteto(0,0,PSIZE,PSIZE,ERASE);		/* eraser bitblit */
	m_func(BIT_SET);
	m_bitwriteto(0,0,PSIZE,PSIZE,PAINT);		/* paint bitblit */
	fname[0] = '\0';
	strcpy(fpwd,getenv("PWD"));

/* go into loop and await events and menu entries */
	while (1) {
		m_gets(buffer);

/* quit signal */
		if (strcmp(buffer,"quit\n") == 0) {
			cleanup();
		}

/* line signals */
		else if (strcmp(buffer,"line\n") == 0) {
			state = LINE;
		}
		else if (strcmp(buffer,"xline\n") == 0) {
			state = XLINE;
		}
		else if (strcmp(buffer,"yline\n") == 0) {
			state = YLINE;
		}
		else if (strcmp(buffer,"box\n") == 0) {
			state = BOX;
		}

		else if (strcmp(buffer,"arrow\n") == 0) {
			sscanf(dbox(100,100,40,2,"Type size of arrow in pixels: "),"%d",&xarrow);
		}

		else if (strcmp(buffer,"circle\n") == 0) {
			state = CIRCLE;
		}

/* erase signal */
		else if (strcmp(buffer,"erase\n") == 0) {
			state = ERASE;
		}

/* paint signal */
		else if (strcmp(buffer,"paint\n") == 0) {
			state = PAINT;
		}

/* brushsize signal */
		else if (strcmp(buffer,"brushsize\n") == 0) {
			sscanf(dbox(100,100,40,2,"Type size of brush in pixels: "),"%d",&xbrush);
			if (xbrush < 1) xbrush = 1;
			if (xbrush > PSIZE) xbrush = PSIZE;
		}

/* text signal */
		else if (strcmp(buffer,"text\n") == 0) {
			state = TEXT;
		}

/* new font signal */
		else if (strcmp(buffer,"newfont\n") == 0) {
			sscanf(dbox(100,100,35,2,"Type new font number: "),"%d",&cfont);
			if (cfont < 0 || cfont >= MAXNFONTS) cfont = 0;
			m_font(cfont);
		}

/* load a file */
		else if (strcmp(buffer,"load\n") == 0) {
			strcpy(fname,dbox(100,100,60,2,"Type file to be loaded: "));
			fnclean(fpwd,fname,fpath);
			if (m_bitfile(1,fpath,&nx,&ny,&nd) == 0) {
				dbox(100,100,60,2,"Can't find it! (Type return to continue) ");
			}
			else {
				m_func(BIT_SRC);
				m_bitcopyto(0,0,nx,ny,0,0,0,1);
			}
		}

/* save a file */
		else if (strcmp(buffer,"save\n") == 0) {
			if (strcmp(fname,"") != 0) {
				sprintf(cline,"File name %s ok? (y/n): ",fname);
				strcpy(buffer,dbox(100,100,60,2,cline));
				if (buffer[0] != 'y') {
					strcpy(buffer,dbox(100,100,60,2,"Type file name: "));
					sscanf(buffer,"%s\n",fname);
				}
				fnclean("/tmp",fname,fpath);
				m_windowsave(fpath);
				move(fpath,fname);
			}
			else {
				strcpy(fname,dbox(100,100,60,2,"Type file name: "));
				fnclean("/tmp",fname,fpath);
				m_windowsave(fpath);
				move(fpath,fname);
			}
/*
			sprintf(cline,"Saved as %s.	Type return to continue: ",fname);
			dbox(100,100,60,2,cline);
*/
		}

/* right button goes down */
		else if (strcmp(buffer,"rbutdown\n") == 0) {

/* set initial line point */
			if (state == LINE || state == XLINE || state == YLINE ||
					state == BOX || state == ERASE || state == CIRCLE) {
				get_mouse(&oldx,&oldy);
				oldx -= OFFSET + xbrush/2;
				oldy -= OFFSET + xbrush/2;
			}

/* paint or erase */
			if (state == PAINT || state == ERASE) {
				m_func(BIT_SRC);
				x = y = -1;
				while (1) {
					oldx = x;
					oldy = y;
					m_getinfo(G_MOUSE2);
					m_gets(buffer);
					if (strcmp(buffer,"rbutup\n") == 0) break;
					sscanf(buffer,"%d %d",&x,&y);
					x -= OFFSET + xbrush/2;
					y -= OFFSET + xbrush/2;
					if (oldx != -1 && oldy != -1 && (x != oldx || y != oldy)) {
						fatline(oldx,oldy,x,y,xbrush,state);
					}
				}
			}
		}

/* right button comes up */
		else if (strcmp(buffer,"rbutup\n") == 0) {

/* draw a line */
			if (state == LINE || state == XLINE || state == YLINE ||
					state == BOX || state == CIRCLE) {
				get_mouse(&x,&y);
				x -= OFFSET + xbrush/2;
				y -= OFFSET + xbrush/2;
				m_func(BIT_SET);
				if (state == LINE) {
					arrowline(oldx,oldy,x,y,xbrush,state,xarrow);
				}
				else if (state == XLINE) {
					arrowline(oldx,oldy,x,oldy,xbrush,state,xarrow);
				}
				else if (state == YLINE) {
					arrowline(oldx,oldy,oldx,y,xbrush,state,xarrow);
				}
				else if (state == BOX) {
					fatline(oldx,oldy,x,oldy,xbrush,state);
					fatline(x,oldy,x,y,xbrush,state);
					fatline(x,y,oldx,y,xbrush,state);
					fatline(oldx,y,oldx,oldy,xbrush,state);
				}
				else if (state == CIRCLE) {
					fdx = x - oldx;
					fdy = y - oldy;
					radius = sqrt(fdx*fdx + fdy*fdy);
					dx = radius;
					dy = 0;
					for (loop = 1; loop <= NCHORDS; loop++) {
						odx = dx;
						ody = dy;
						dangle = 360./NCHORDS;
						angle = loop*(dangle/RADIAN);
						dx = radius*cos(angle);
						dy = radius*sin(angle);
						fatline(oldx + odx,oldy + ody,oldx + dx,oldy + dy,xbrush,state);
					}
				}
			}

/* write some text */
			else if (state == TEXT) {
				get_mouse(&x,&y);
				m_movecursor(x - XTOFF,y - YTOFF);
				m_setcursor(CS_BLOCK);
				m_setecho();
				m_func(BIT_SET);
				fgets(buffer,LINESIZE,stdin);
				m_setcursor(CS_INVIS);
				m_setnoecho();
			}
		}
	}
}

/* dialog box */
char *dbox(x,y,dx,dy,query)
int x,y;			/* position of box */
int dx,dy;			/* size of box */
char *query;
{
char barf[40];
int window;
int fontx,fonty;
int winlocx,winlocy,winsizex,winsizey;

	m_font(0);
        m_getfontsize(&fontx,&fonty);
        m_getwindowposition(&winlocx,&winlocy);
        m_getwindowsize(&winsizex,&winsizey);
	window = m_makewindow(winlocx + x,winlocy + y,dx*fontx,dy*fonty);
	m_selectwin(window);
	m_setecho();
	printf("%s",query);
	m_gets(buffer);
	m_setnoecho();
	m_destroywin(window);
	m_gets(barf);
	m_font(cfont);
	return(buffer);
}

/* clean up file name and add path if needed */
void
fnclean(pwd,name,path)
char *pwd,*name,*path;
{
int i;

	for (i = 0; name[i] != '\0'; i++) {
		if (name[i] == '\n') {
			name[i] = '\0';
			break;
		}
	}
	if (name[0] != '/') sprintf(path,"%s/%s",pwd,name);
	else strcpy(path,name);
}

/* make a line with an arrow */
void
arrowline(oldx,oldy,x,y,xbrush,state,xarrow)
int oldx,oldy;
int x,y;
int xbrush;
int state;
int xarrow;
{
int nx,ny,px,py,len;

	fatline(oldx,oldy,x,y,xbrush,state);
	if (xarrow > 0) {
		nx = oldx - x;
		ny = oldy - y;
		len = abs(nx) + abs(ny);
		if (len > 0) {
			nx *= xarrow;
			ny *= xarrow;
			nx /= len;
			ny /= len;
			px = nx - ny/2 + x;
			py = nx/2 + ny + y;
			fatline(x,y,px,py,xbrush,state);
			px = nx + ny/2 + x;
			py = -nx/2 + ny + y;
			fatline(x,y,px,py,xbrush,state);
		}
	}
}

/* make a fat line using bit blit call -- otherwise use line call */
void
fatline(oldx,oldy,x,y,xbrush,state)
int oldx,oldy;
int x,y;
int xbrush;
int state;
{
int i,j,k;
float fract;

	if (xbrush == 1) {
		m_line(oldx,oldy,x,y);
	}
	else {
		if (abs(x - oldx) > abs(y - oldy)) {
			j = x > oldx ? 1 : -1;
			for (i = oldx; i != x; i += j) {
				if (x == oldx) fract = 0.;
				else fract = (float)(i - oldx)/(float)(x - oldx);
				if (y == oldy) k = y;
				else k = (fract*y + (1. - fract)*oldy) + .5;
				m_bitcopyto(i,k,xbrush,xbrush,0,0,0,state);
			}
		}
		else {
			j = y > oldy ? 1 : -1;
			for (i = oldy; i != y; i += j) {
				if (y == oldy) fract = 0.;
				else fract = (float)(i - oldy)/(float)(y - oldy);
				if (x == oldx) k = x;
				else k = (fract*x + (1. - fract)*oldx) + .5;
				m_bitcopyto(k,i,xbrush,xbrush,0,0,0,state);
			}
		}
	}
}

/* cleanup routine */
void cleanup()
{
	m_clearmenu(1);
	m_font(0);
	m_clear();
	m_pop();
	m_clearmode(M_ABS);
	m_setecho();
	m_setcursor(CS_BLOCK);
	exit(0);
}

/* move -- move a file from one place to another */
void
move(src,dst)
char *src,*dst;
{
	char tstr[512];
	if (strcmp(src,dst) != 0) {
		sprintf(tstr,"sleep 2;mv %s %s",src,dst);
		system(tstr);
	}
}
